lib/chef/resource_resolver.rb

```diff
@@ -98,59 +99,70 @@ def resolve
 
     # @api private
     def list
-      Chef::Log.debug "Resources for generic #{resource_name} resource enabled on node include: #{prioritized_handlers}"
-      prioritized_handlers
+      Chef::Log.debug "Resources for generic #{resource_name} resource enabled on node include: #{enabled_handlers}"
+      enabled_handlers
     end
 
     #
     # Whether this DSL is provided by the given resource_class.
     #
+    # Does NOT call provides? on the resource (it is assumed this is being
+    # called *from* provides?).
+    #
     # @api private
     def provided_by?(resource_class)
-      !prioritized_handlers.include?(resource_class)
+      potential_handlers.include?(resource_class)
+    end
+
+    #
+    # Whether the given handler attempts to provide the resource class at all.
+    #
+    # @api private
+    def self.includes_handler?(resource_name, resource_class)
+      priority_map.includes_handler?(resource_name, resource_class)
     end
 
     protected
 
+    def self.priority_map
+      Chef::Platform::ResourcePriorityMap.instance
+    end
+
     def priority_map
       Chef::Platform::ResourcePriorityMap.instance
     end
 
-    def prioritized_handlers
-      @prioritized_handlers ||=
-        priority_map.list_handlers(node, resource_name, canonical: canonical)
+    # @api private
+    def potential_handlers
+      priority_map.list_handlers(node, resource_name, canonical: canonical).flatten(1).uniq
```

Comment:
Given that this `.flatten(1).uniq` exists in both places that use `list_handlers` and that expression is suuuuupre critical to the functioning of this code, maybe it should be folded in to `list_handlers` itself? At a minimum there should be some comments that changing the format requires updating all three places.